Services/WatchlistService.swift

```diff
@@ -6,76 +6,129 @@
+        } catch {
+            print("Fetch failed")
+            throw WatchlistServiceError.failedToFetchFromPersistentStore
         }
-
-        return items
     }
     
-    private func getWatchlist() -> Watchlist {
-        let defaults = UserDefaults.standard
+    func toggleStatus(for item: WatchlistMovieConfiguration) -> WatchlistStatus {
+        let request = WatchlistMovie.createFetchRequest()
+        predicate = NSComparisonPredicate(format: "id == %@", item.id as NSNumber)
+        request.predicate = predicate
         
-        if let watchlistData = defaults.object(forKey: WatchlistService.watchlistKey) as? Data {
-            let decoder = JSONDecoder()
-            if let watchlist = try? decoder.decode(Watchlist.self, from: watchlistData) {
-                return watchlist
+        do {
+            let movies = try container.viewContext.fetch(request)
+            
+            if !movies.isEmpty {
+                container.viewContext.delete(movies[0])
+                return saveWatchlist() ? .notAdded : .added
             } else {
-                return Watchlist()
+                let watchlistMovie = WatchlistMovie(context: container.viewContext)
+                watchlistMovie.id = item.id
+                watchlistMovie.saveDate = item.saveDate
+                watchlistMovie.title = item.title
+                watchlistMovie.voteAverage = item.voteAverage
+                watchlistMovie.posterPath = item.posterPath
+
+                return saveWatchlist() ? .added : .notAdded
             }
-        } else {
-            return Watchlist()
+        } catch {
+//            replace with throwing function later
+            return saveWatchlist() ? .added : .notAdded
         }
+        
     }
     
-    private func saveWatchlist(_ list: Watchlist) -> Bool {
-        let encoder = JSONEncoder()
+    func getWatchlist() -> [WatchlistMovie] {
+        let request = WatchlistMovie.createFetchRequest()
+        let sort = NSSortDescriptor(key: "saveDate", ascending: false)
+        request.sortDescriptors = [sort]
         
-        if let watchlistData = try? encoder.encode(list) {
-            let defaults = UserDefaults.standard
-            defaults.set(watchlistData, forKey: WatchlistService.watchlistKey)
-            return true
-        } else {
-            return false
+        do {
+            let watchlistMovies = try container.viewContext.fetch(request)
+            return watchlistMovies
+        } catch {
+            print("Fetch failed")
+            return [WatchlistMovie]()
         }
     }
+    
+    private func saveWatchlist() -> Bool {
+        if container.viewContext.hasChanges {
+                do {
+                    try container.viewContext.save()
+                    return true
+                } catch {
+                    print("An error occurred while saving: \(error)")
+                    return false
+                }
+        }
+        
+        return true
```

Comment:
Since `hasChanges` is a required step for this method, making it a `guard` improves readability
```suggestion
        guard container.viewContext.hasChanges else { return true }
        
        do {
            try container.viewContext.save()
            return true
        catch {
            print("An error occurred while saving: \(error)")
            return false
        }
```